### PR TITLE
API keys cleanup

### PR DESCRIPTION
**Pull request by aleksmaus**
## What is the problem this PR solves?

Adds the API keys cleanup/invalidation.

## How does this PR solve the problem?

1. If the API key needs to be updated, the old API keys are appended to the .fleet-agents ```default_api_key_history``` property.
2. When the policy ack is received the keys from the history are invalidated and the agent document ```default_api_key_history``` is reset

## How to test this PR locally

1. Enroll the agent, check the ```default_api_key``` and ```default_api_key_id```
2. Update the policy, by adding new integration. The permissions change will cause the API key to be regenerated and the old keys to be appended to the history. Should be able to observer the ```default_api_key_history``` updated with the previous key before the policy change ack is recieved:

Screen Shot 2021-11-16 at 4 05 14 PM

3. When the policy ack is received, should be able to see in the fleet server log, the ```"Invalidate old API keys"``` message.

Screen Shot 2021-11-16 at 4 06 13 PM

4. Check the agent record again after that, confirm that the history is cleared.

Screen Shot 2021-11-16 at 4 08 37 PM

5. Check the API keys
```
GET /_security/api_key
```
Confirm that the old key id invalidated:
Screen Shot 2021-11-16 at 4 07 38 PM

@scunningham there are still probably some edge cases, but this should improve the situation.
for example:
* what happens if the agent never acks?
* how should we handle invalidation errors? 
* what if only some keys fail to invalidate? how many time should we retry? or how long should we keep the key in the history?

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by aleksmaus** [comment]
Once we agree on the schema change I'll open PR to elasticsearch to update the agent mapping.

**Comment by ruflin** [comment]
@aleksmaus Are we talking here about the API which communicates to fleet-server or the ES shipping API key or both?

**Comment by aleksmaus** [comment]
> @aleksmaus Are we talking here about the API which communicates to fleet-server or the ES shipping API key or both?

this is for elasticsearch API keys, that we regenerate on each policy permissions change

**Comment by ruflin** [comment]
I like the direction of this approach. Having the information persisted in Elasticsearch around the status of the API keys and some history also allows potential future cleanup for error/edge cases and also gives Fleet visibility into it. This means assuming a user in Fleet would have enough permissions, we could also show mismatches on the UI (@joshdover ).

File: cmd/fleet/handleAck.go [hunk 1]
```diff
@@ -360,7 +372,7 @@ func _getAPIKeyIDs(agent *model.Agent) []string {
 //
 // WARNING: This assumes the input data is sanitized.
 
-const kUpdatePolicyPrefix = `{"script":{"lang":"painless","source":"if (ctx._source.policy_id == params.id) {ctx._source.` +
+const kUpdatePolicyPrefix = `{"script":{"lang":"painless","source":"if (ctx._source.policy_id == params.id) {ctx._source.remove('default_api_key_history');ctx._source.` +
```
Comment (scunningham) [inline cmd/fleet/handleAck.go:375]: I'm sorry you had to see this.  Look away.

File: cmd/fleet/handleCheckin.go [hunk 2]
```diff
@@ -476,10 +478,17 @@ func processPolicy(ctx context.Context, zlog zerolog.Logger, bulker bulk.Bulk, a
 			dl.FieldDefaultApiKeyId:             defaultOutputApiKey.Id,
 			dl.FieldPolicyOutputPermissionsHash: pp.Default.Role.Sha2,
 		}
+		if agent.DefaultApiKeyId != "" {
+			fields[dl.FieldDefaultApiKeyHistory] = model.DefaultApiKeyHistoryItems{
+				ApiKey:    agent.DefaultApiKey,
```
Comment (scunningham) [inline cmd/fleet/handleCheckin.go:483]: Do we need to keep the old key material in the history?

**Review by scunningham — APPROVED** [review]

File: cmd/fleet/handleCheckin.go (hunk 2 above)
Comment (aleksmaus) [reply cmd/fleet/handleCheckin.go:483]: > Do we need to keep the old key material in the history?

Not really at the moment. Just added all the default_api_key related properties there. Can remove, let me know.

Comment (scunningham) [reply cmd/fleet/handleCheckin.go:483]: Fewer secrets lying around the better I think.

Comment (aleksmaus) [reply cmd/fleet/handleCheckin.go]: Done.
Screen Shot 2021-11-17 at 10 20 16 AM

**Comment by scunningham** [comment]
👍

File: cmd/fleet/handleCheckin.go (hunk 2 above)
Comment (aleksmaus) [reply cmd/fleet/handleCheckin.go]: image

File: model/schema.json [hunk 3]
```diff
@@ -442,6 +442,24 @@
           "description": "API key the Elastic Agent uses to authenticate with elasticsearch",
           "type": "string"
         },
+        "default_api_key_history": { 
+          "description": "Default API Key History",
+          "type": "array",
```
Comment (ruflin) [inline model/schema.json:447]: Could we maybe set this to index: false instead? https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-index.html The potential problem I see with this structure if we index is that each entry creates a separate document in lucene which creates unnecessary overhead if we never have to query the data.

Comment (aleksmaus) [reply model/schema.json:447]: yep. will have that set to false for elasticsearch PR for .fleet-agents mapping.

**Pull request merged**